### DISCUSSION TITLE
Pre-processor for rules

### DISCUSSION THREAD

**Discussion by vasily-kartashov** [tuProlog/2p-kt#11] [Q&A] [answered]
When loading a file with an example rule:

```prolog
:- response_body(endpoint, [json]).
endpoint([message('Hello, World!')]).
```

I'd like to automatically replace it with the following rule (based on the presence of `response_body` "annotation"):

```prolog
endpoint(Response) :-
     original_endpoint(Response0), // keeping the original goal
     term_json(Response0, Payload),
     Response = [headers(['Content-Type': 'application/json']), payload(Payload)].
```

What is the best way to achieve it?

**Comment by gciatto** [comment, marked as answer]
The first line you wrote its a directive, which is not properly a rule.

Both rules and directives are clauses.

A theory is a sequence of clauses, essentially.

So the best way to do what you propose is to parse the theory, selecting the directive and manipulating it from java, thus crating a new theory having a rule instead of a directory

**Reply by vasily-kartashov** [reply]
@gciatto-unibo It seems to me this is quite a bit more complex than I thought it would be.

Given for example this predicate:

```prolog
len([], 0).
len([_|T], I) :-
    len(T, J),
    I is J + 1.
```

the rewrite rule should recursively replace all `len` with `len_internal` and only then add the wrapper

```
len_internal([], 0).
len_intenal([_|T], I) :-
    len_internal(T, J),
    I is J + 1.
    
len(A, B): -
   % here comes the boilerplate
   len_internal(A, B).
```

Adding the fact that Clauses are immutable, transforming the trees in Java seems too complex. 

Does tuProlog support term expansion? Can I transform a theory with Prolog before sending it to solver?

**Reply by gciatto** [reply]
What do you mean by "term expansion"?

However, I don't know what you have tried to perform such a kind of transformation, but here are my 50 cents.

The transformation involves 
1. the renaming of some predicates occurring into the theory and 
2. the addition of some clauses.

Step 2 is straightforward, given the `assertZ` method of theories.
Step 1 is what is giving you troubles, I bet.

The immutable design of 2P's terms and classes is intended, and the same is true for the difficulty in modifying terms and clauses. The preferred way to do so is to rewrite terms for scratch, possibly while varying something, e.g. a functor.

The best way to apply a whole transformation to a term/clause is to exploit the notion of `TermVisitor`.
Here's an example: https://github.com/tuProlog/2p-kt/blob/feature/examples/examples/src/test/java/it/unibo/tuprolog/examples/TermVisitorExamplesJvm.java

Essentially one must create a visitor aimed at renaming a single clause if it matches a particular patter, or leave the clause unaffected otherwise. Then, one must apply such visitor to each clause in the original theory via the `Term.accept` method.
This should apply the renaming to each clause.
The new list of clauses should then be used to create a new theory, to which further clauses may be optionally added.

**Reply by vasily-kartashov** [reply]
The visitor pattern makes perfect sense, I was hoping there is something like this. Thank you.

FYI, this is `term_expansion` predicate from SWI: https://www.swi-prolog.org/pldoc/doc_for?object=term_expansion/2

**Reply by gciatto** [reply]
Yeah the visitor pattern is cool, we use it a lot behind the scenes to perform such kinds of manipulations :)

No, `term_expansion` is not supported ATM.

**Reply by vasily-kartashov** [reply]
@gciatto-unibo Thank you again for the example! So far I find the overall API design logical and consistent and your help is very valuable.

**Reply by gciatto** [reply]
Thank you mate. Very appreciated.

Thank you too for giving me the opportunity of writing these examples
